crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space/bootstrap_task.rs

```diff
@@ -0,0 +1,708 @@
+        assert!(
+            // It's 15 not 22 because the task actually slept for 1 + 3 + 9 milliseconds and we are reading the delay
+            // after it has been updated.
+            start_time.elapsed() >= Duration::from_millis(15),
+            "Bootstrap task should have slept for at least as long as the delay values we saw"
+        );
+
+        test_sender.ghost_actor_shutdown_immediate().await.unwrap();
+    }
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn bootstrap_task_prevents_sleep_disable_via_multiplier() {
+        let agents = vec![fixt!(AgentInfoSigned)];
+        let (test_sender, mut host_stub, task) = setup(
+            DummySpaceInternalImpl::new(HashSet::new()),
+            agents,
+            // Set to 0 to try and force skipping the sleep
+            0,
+            false,
+        )
+        .await;
+
+        let (mut sender, receiver) = channel(3);
+        tokio::time::timeout(Duration::from_secs(30), {
+            async move {
+                for _ in 0..3 {
+                    host_stub.next_event(Duration::from_secs(5)).await;
+                    sender.send(task.read().current_delay).await.unwrap();
+                }
+            }
+        })
+        .await
+        .unwrap();
+
+        let durations = receiver.map(|d| d.as_millis()).collect::<Vec<u128>>().await;
+
+        assert_eq!(
+            vec![2, 4, 8],
+            durations,
+            "Expected durations to increase exponentially"
+        );
+
+        test_sender.ghost_actor_shutdown_immediate().await.unwrap();
+    }
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn bootstrap_task_query_delay_increase_respects_max() {
+        let agents = vec![fixt!(AgentInfoSigned)];
+        // Set a high delay multiplier so that a multiplication with no max check would move the delay to 1s from 1ms,
+        // instead of the actual max at 10ms.
+        let (test_sender, mut host_stub, task) = setup(
+            DummySpaceInternalImpl::new(HashSet::new()),
+            agents,
+            1000,
+            false,
+        )
+        .await;
+
+        let (mut sender, receiver) = channel(3);
+        tokio::time::timeout(Duration::from_secs(30), {
+            async move {
+                for _ in 0..3 {
+                    host_stub.next_event(Duration::from_secs(5)).await;
+                    sender.send(task.read().current_delay).await.unwrap();
+                }
+            }
+        })
+        .await
+        .unwrap();
+
+        let durations = receiver.map(|d| d.as_millis()).collect::<Vec<u128>>().await;
+
+        assert_eq!(
+            vec![10, 10, 10],
+            durations,
+            "Expected durations to increase exponentially"
+        );
+
+        test_sender.ghost_actor_shutdown_immediate().await.unwrap();
+    }
+
+    // TODO should this be trusted? It's an external call
+    #[tokio::test(flavor = "multi_thread")]
+    async fn bootstrap_task_query_agent_limit_is_trusted() {
```

Comment:
There's a potential security question here. We set a limit of 8 when querying for agents but we never actually check the bootstrap server respected this. Paired with the comment in the existing code about validating the bootstrap response I think this would be worth a bit more effort to look at

Reply:
I'm thinking the limit not being respected is a sign that the remote is misbehaving by not respecting its own API. My preference here would be to treat this as untrusted input and if there is a hint that the remote is doing something it shouldn't, like giving us 1,000 agents when we expected a max of 8, then we should reject the response and log an error. Adding a questionable response to our database means we will go ahead and gossip with those agents.

I think as a precaution this should be checked and logged, it doesn't run super often so it's not going to create too much noise. If the remote is misbehaving we won't even know at the moment except by looking at the agent store.

Reply:
Yep, that makes sense. I'm fine with that approach.

Reply:
Awesome okay, I think I'll take this on with a follow up PR to not overdo the number of changes on one PR